Client tools auto update

vapopov opened the pull request:
In this PR implemented logic for client tools auto update {tsh, tctl} described in this RFD https://github.com/gravitational/teleport/pull/39805

Update logic downloads Teleport binaries from CDN, extracts data in tools sub-directory (going to be used `$TELEPORT_HOME/bin/{UUID-update-pkg}/{tsh,tctl}`) and creates symlinks for {tsh,tctl} to `$TELEPORT_HOME/bin/{tsh,tctl}`.
Added integration tests which compiles updater binary with update logic only to simulate command execution with requested version to downloads them from fake test CDN server.

vapopov committed:
Add client tools auto update

vapopov commented on lib/utils/disk.go:
```diff
@@ -53,10 +53,7 @@ func FreeDiskWithReserve(dir string, reservedFreeDisk uint64) (uint64, error) {
-		return 0, trace.Errorf("invalid size")
-	}
-	avail := stat.Bavail * uint64(stat.Bsize)
+	avail := uint64(stat.Bavail) * uint64(stat.Bsize)
```
This change related to this comment https://github.com/gravitational/teleport/pull/47060#discussion_r1795868236
also I'm locally checking linter for warnings related to`uint64(stat.Bavail)`

sclevine commented on lib/autoupdate/progress.go:
```diff
@@ -0,0 +1,81 @@
+
+// cancelableTeeReader is a copy of TeeReader with ability to react on signal notifier
+// to cancel reading process.
+func cancelableTeeReader(r io.Reader, w io.Writer, signals ...os.Signal) io.Reader {
```
It's safe to call Close while concurrently Reading from a response body. Instead of creating this custom Reader, I'd recommend something like this:
```golang
go func() {
   <-ctx.Done()
   resp.Body.Close()
}()
```

sclevine replied:
Actually, NewRequestWithContext should cover reading the response body. Have you tested without using cancelableTeeReader? Seems like it should just work.

vapopov replied:
I've replaced to `NewRequestWithContext` recently, let me check

sclevine commented on lib/autoupdate/utils.go:
```diff
@@ -0,0 +1,168 @@
+	return filepath.Join(filepath.Clean(home), ".tsh", "bin"), nil
+}
+
+func version(toolsDir string) (string, error) {
```
Recommend longer global function name that won't be shadowed by variables called `version`.

sclevine commented on lib/autoupdate/client_update.go:
```diff
@@ -0,0 +1,353 @@
+
+	h := sha256.New()
+	pw := &progressWriter{n: 0, limit: resp.ContentLength}
+	body := cancelableTeeReader(io.TeeReader(resp.Body, h), pw, syscall.SIGINT, syscall.SIGTERM)
```
> context should be canceled by signal anyway if we want to stop downloading

Exactly, better to use the context abstraction, instead of hooking into the OS at this level

vapopov committed:
Replace fork for posix platform for re-exec
Move integration tests to client tools specific dir
Use context cancellation with SIGTERM, SIGINT
Remove cancelable tee reader with context replacement
Renaming

vapopov committed:
Fix syscall path execution
Fix archive cleanup if hash is not valid
Limit the archive write bytes

vapopov committed:
Merge remote-tracking branch 'origin/master' into vapopov/client-auto-update

vapopov committed:
Cover the case with single package for darwin platform after v17

vapopov committed:
Move updater logic to tools package

sclevine commented on lib/autoupdate/tools/utils.go:
```diff
@@ -0,0 +1,185 @@
+
+// teleportPackageURLs returns the URL for the Teleport archive to download. The format is:
+// https://cdn.teleport.dev/teleport-{, ent-}v15.3.0-{linux, darwin, windows}-{amd64,arm64,arm,386}-{fips-}bin.tar.gz
+func teleportPackageURLs(baseUrl, toolsVersion string) ([]packageURL, error) {
```
```suggestion
func teleportPackageURLs(baseURL, toolsVersion string) ([]packageURL, error) {
```

vapopov replied:
my bad, I've renamed all occurrences

sclevine commented on integration/autoupdate/tools/main_test.go:
```diff
@@ -0,0 +1,193 @@
+		"-ldflags", strings.Join([]string{
+			fmt.Sprintf("-X 'main.toolsDir=%s'", toolsDir),
+			fmt.Sprintf("-X 'main.version=%s'", version),
+			fmt.Sprintf("-X 'main.baseUrl=http://%s'", baseURL),
```
```suggestion
			fmt.Sprintf("-X 'main.baseURL=http://%s'", baseURL),
```

sclevine commented on lib/autoupdate/tools/updater.go:
```diff
@@ -0,0 +1,390 @@
+	localVersion string
+	tools        []string
+
+	baseUrl string
```
```suggestion
	baseURL string
```

sclevine commented on lib/autoupdate/tools/updater.go:
```diff
@@ -0,0 +1,390 @@
+	client  *http.Client
+}
+
+// NewUpdater initiate updater for the client tools auto updates.
```
Can you document what each of the arguments takes? E.g., what format should be used for `tools`?

sclevine commented on lib/autoupdate/tools/updater.go:
```diff
@@ -0,0 +1,390 @@
+		return trace.Wrap(err)
+	}
+
+	signalCtx, cancel := signal.NotifyContext(ctx, syscall.SIGINT, syscall.SIGTERM)
```
Will the original context get canceled anyways if we receive these signals?

tsh/tctl/tbot will have their own signal handling, so it seems better to drive this via the original context.

vapopov replied:
as far as I remember there were cases where we have no context propagated. I removed this part from library updater logic, will handle with tctl and tsh integration

sclevine commented on lib/autoupdate/tools/updater.go:
```diff
@@ -0,0 +1,390 @@
+	}
+
+	var buf bytes.Buffer
+	_, err = io.CopyN(&buf, resp.Body, sha256.Size*2)
```
```suggestion
	_, err = io.CopyN(&buf, resp.Body, sha256.Size*2) // SHA bytes to hex
```

vapopov committed:
Move context out from the library
Base URL renaming

vapopov committed:
Add more context in comments

vapopov committed:
Merge remote-tracking branch 'origin/master' into vapopov/client-auto-update

vapopov committed:
Changes in find endpoint

sclevine commented on lib/autoupdate/tools/utils.go:
```diff
@@ -0,0 +1,185 @@
+
+const (
+	// FlagEnt represents enterprise version.
+	FlagEnt = 1 << 0
```
```suggestion
	FlagEnt = 1 << iota
```
Then you can avoid repeating the expression.

(also might consider unexporting these, since they are only used internally)

sclevine commented on lib/autoupdate/tools/updater.go:
```diff
@@ -0,0 +1,391 @@
+func (u *Updater) CheckLocal() (string, bool) {
+	// Check if the user has requested a specific version of client tools.
+	requestedVersion := os.Getenv(teleportToolsVersionEnv)
+	switch {
+	// The user has turned off any form of automatic updates.
+	case requestedVersion == "off":
+		return "", false
+	// Requested version already the same as client version.
+	case u.localVersion == requestedVersion:
+		return requestedVersion, false
+	}
```
```suggestion
	switch requestedVersion {
	// The user has turned off any form of automatic updates.
	case "off":
		return "", false
	// Requested version already the same as client version.
	case u.localVersion:
		return requestedVersion, false
	}
```

sclevine commented on lib/autoupdate/tools/updater.go:
```diff
@@ -0,0 +1,391 @@
+}
+
+// CheckLocal is run at client tool startup and will only perform local checks.
+func (u *Updater) CheckLocal() (string, bool) {
```
Document return params

sclevine commented on lib/autoupdate/tools/updater.go:
```diff
@@ -0,0 +1,391 @@
+
+// CheckRemote will check against Proxy Service if client tools need to be
+// updated.
+func (u *Updater) CheckRemote(ctx context.Context, proxyAddr string) (string, bool, error) {
```
Document return params

sclevine replied:
Still not clear to me what the bool means by reading the doc. Could you explain that in the godoc?
Might also help to name the params:
```
(version string, changed bool, er error)
```

sclevine commented on lib/autoupdate/tools/updater.go:
```diff
@@ -0,0 +1,391 @@
+func (u *Updater) CheckRemote(ctx context.Context, proxyAddr string) (string, bool, error) {
+	// Check if the user has requested a specific version of client tools.
+	requestedVersion := os.Getenv(teleportToolsVersionEnv)
+	switch {
+	// The user has turned off any form of automatic updates.
+	case requestedVersion == "off":
+		return "", false, nil
+	// Requested version already the same as client version.
+	case u.localVersion == requestedVersion:
+		return requestedVersion, false, nil
+	}
```
```suggestion
	switch requestedVersion {
	// The user has turned off any form of automatic updates.
	case "off":
		return "", false, nil
	// Requested version already the same as client version.
	case u.localVersion:
		return requestedVersion, false, nil
	}
```

sclevine commented on lib/autoupdate/tools/updater.go:
```diff
@@ -0,0 +1,391 @@
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	raw := buf.String()
+	if _, err = hex.DecodeString(raw); err != nil {
+		return "", trace.Wrap(err)
+	}
+	return raw, nil
```
Suggest returning the decoded hash and comparing the actual bytes, instead of comparing hex (which has multiple valid representations)

sclevine commented on lib/autoupdate/tools/updater.go:
```diff
@@ -0,0 +1,391 @@
+	return raw, nil
+}
+
+func (u *Updater) downloadArchive(ctx context.Context, downloadDir string, url string) (string, string, error) {
```
Document return params

sclevine commented on lib/autoupdate/tools/updater.go:
```diff
@@ -0,0 +1,391 @@
+
+	// Caller of this function will remove this file after the atomic swap has
+	// occurred.
+	f, err := os.CreateTemp(downloadDir, "tmp-")
```
This file descriptor is leaked. It's impossible to close after the method returns, and writing may not actually be complete. This needs a close with an error check.

Also, I'd suggest passing the file path (or maybe just a Writer) to write to as an argument, so that the caller knows that it must delete the file if the method returns an error, and additionally, deleting the file on error.

Example: https://github.com/gravitational/teleport/pull/47565/files#diff-8961b5cac012637602dcc182a8b40b169a527f663072dc34b0ab2db12922e86dR118-R127

vapopov committed:
Replace test http server with `httptest`
Replace hash for bytes matching
Proper temp file close for archive download

vapopov committed:
Add more context to comments

vapopov committed:
Move feature flag to main package to be reused

sclevine commented on lib/autoupdate/feature.go:
```diff
@@ -0,0 +1,38 @@
+	// FlagEnt represents enterprise version.
+	FlagEnt = 1 << iota
+	// FlagFips represents enterprise version with fips feature enabled.
+	FlagFips
```
```suggestion
	FlagFIPS
```

sclevine commented on lib/autoupdate/tools/utils.go:
```diff
@@ -0,0 +1,174 @@
+		}
+	}
+
+	return filepath.Join(filepath.Clean(home), ".tsh", "bin"), nil
```
Note that filepath.Join runs filepath.Clean on the result for you

vapopov committed:
Constant rename

hugoShaka commented on lib/autoupdate/feature_ent.go:
```diff
@@ -0,0 +1,26 @@
+//go:build webassets_ent
+// +build webassets_ent
```
I'm not sure if this is a reliable way to detect ent builds. When developping we often do ent builds without webassets.

vapopov replied:
this one was wrong way to identify the build version, replacing with `lib/modules`

hugoShaka reviewed (approved):
Looks good to me ~but we might wanbt to check if there's a saner way to detect enterprise builds. I'll open a thread in #teleport-dev~

Edit: just saw your comment

sclevine commented:
@hugoShaka note lib/modules might not work for tbot: https://github.com/gravitational/teleport/pull/47565#discussion_r1806888606

vapopov committed:
Replace build tag with lib/modules to identify enterprise build

vapopov commented on lib/autoupdate/fips_enabled.go:
```diff
@@ -0,0 +1,25 @@
+//go:build fips
```
@zmb3 should I move this one to lib/modules to identify if the build supports the FIPS? Or there already a way to get this info from modules package

sclevine replied:
I think this `modules.GetModules().IsBoringCrypto()` -- BoringCrypto is FIPS validated

hugoShaka commented:
> @hugoShaka note lib/modules might not work for tbot: [#47565 (comment)](https://github.com/gravitational/teleport/pull/47565#discussion_r1806888606)

I think tbot should be considered an agent like teleport and subject to updates from the agent updater we're building. According to the RFD we going to have a single agent updater for oss, ent, fips and it will figure out the right build from /find or /ping so I think everything will fit nicely?

vapopov committed:
Replace fips tag with modules flag

sclevine commented:
@hugoShaka tbot also has a oneshot mode for CI/CD use cases. The current (possibly only verbally discussed) plan is to allow tbot to be upgraded through either method. But this may not ship in the first release of client tools updates.

vapopov commented:
@avatus @EdwardDowling would appreciate your review, need one more approval
